hazelcast/protocol/builtin.py

```diff
@@ -496,3 +554,206 @@ def encode(buf, value, is_final=False):
+        return ListCNFixedSizeCodec.decode(msg, BYTE_SIZE_IN_BYTES, FixSizedTypesCodec.decode_byte)
+
+
+class ListCNShortCodec(object):
+    @staticmethod
+    def decode(msg):
+        return ListCNFixedSizeCodec.decode(
+            msg, SHORT_SIZE_IN_BYTES, FixSizedTypesCodec.decode_short
+        )
+
+
+class ListCNIntegerCodec(object):
+    @staticmethod
+    def decode(msg):
+        return ListCNFixedSizeCodec.decode(msg, INT_SIZE_IN_BYTES, FixSizedTypesCodec.decode_int)
+
+
+class ListCNLongCodec(object):
+    @staticmethod
+    def decode(msg):
+        return ListCNFixedSizeCodec.decode(msg, LONG_SIZE_IN_BYTES, FixSizedTypesCodec.decode_long)
+
+
+class ListCNFloatCodec(object):
+    @staticmethod
+    def decode(msg):
+        return ListCNFixedSizeCodec.decode(
+            msg, FLOAT_SIZE_IN_BYTES, FixSizedTypesCodec.decode_float
+        )
+
+
+class ListCNDoubleCodec(object):
+    @staticmethod
+    def decode(msg):
+        return ListCNFixedSizeCodec.decode(
+            msg, DOUBLE_SIZE_IN_BYTES, FixSizedTypesCodec.decode_double
+        )
+
+
+class ListCNLocalDateCodec(object):
+    @staticmethod
+    def decode(msg):
+        return ListCNFixedSizeCodec.decode(
+            msg, _LOCAL_DATE_SIZE_IN_BYTES, FixSizedTypesCodec.decode_local_date
+        )
+
+
+class ListCNLocalTimeCodec(object):
+    @staticmethod
+    def decode(msg):
+        return ListCNFixedSizeCodec.decode(
+            msg, _LOCAL_TIME_SIZE_IN_BYTES, FixSizedTypesCodec.decode_local_time
+        )
+
+
+class ListCNLocalDateTimeCodec(object):
+    @staticmethod
+    def decode(msg):
+        return ListCNFixedSizeCodec.decode(
+            msg, _LOCAL_DATE_TIME_SIZE_IN_BYTES, FixSizedTypesCodec.decode_local_date_time
+        )
+
+
+class ListCNOffsetDateTimeCodec(object):
+    @staticmethod
+    def decode(msg):
+        return ListCNFixedSizeCodec.decode(
+            msg, _OFFSET_DATE_TIME_SIZE_IN_BYTES, FixSizedTypesCodec.decode_offset_date_time
+        )
+
+
+class BigDecimalCodec(object):
+    @staticmethod
+    def decode(msg):
+        buf = msg.next_frame().buf
+        size = FixSizedTypesCodec.decode_int(buf, 0)
+        unscaled_value = int_from_bytes(buf[INT_SIZE_IN_BYTES : INT_SIZE_IN_BYTES + size])
+        scale = FixSizedTypesCodec.decode_int(buf, INT_SIZE_IN_BYTES + size)
+        sign = 0 if unscaled_value >= 0 else 1
+        return Decimal((sign, tuple(map(int, str(abs(unscaled_value)))), -1 * scale))
+
+
+class SqlPageCodec(object):
+    @staticmethod
+    def decode(msg):
+        from hazelcast.sql import SqlColumnType, _SqlPage
+
+        # begin frame
+        msg.next_frame()
+
+        # read the "last" flag
+        is_last = LE_INT8.unpack_from(msg.next_frame().buf, 0)[0] == 1
+
+        # read column types
+        column_type_ids = ListIntegerCodec.decode(msg)
+
+        # read columns
+        columns = []
+
+        for column_type_id in column_type_ids:
+            if column_type_id == SqlColumnType.VARCHAR:
```

Comment:
I tried this but the end result was something uglier (to my taste :smile: )

Would it be okay to leave it like this?

Reply:
Well, apart from being idiomatic, `dict` approach is also faster. Long if/elif chains (and loops) are slow in Python, but `dict` is implemented in C. In the code below, there's quite a difference between both approaches: https://gist.github.com/yuce/cadd9b39a06bee679b75454004d520aa

Using the `dict` approach also makes it possible to get rid of the for loop and use a list comprehension which would make it even faster.

Reply:
I don't think performance should be a decisive factor here. 

If you change the gist you shared with the code I pasted there, I couldn't see much difference. In the code I shared, I passed 10 as the column type (since on average we should expect to make (1 + 19)/2 comparisons) and did something pretty simple to decode a buffer. 

Note that, on the actual code, the overhead is much larger (decoding the whole row-set for a single column per column type), I believe dict or if/else approach would be identical because the actual decoding would dwarf the iteration method overhead

Reply:
OK, fair enough. Here's how the dict approach looks like, it helps getting rid of the loop (IMO in an elegant way):
```python
# read columns
try:
    columns = [_DECODER[tid](msg) for tid in column_type_ids]
except KeyError as e:
    raise ValueError("Unknown type %s" % e.args[0])
# ...
```
There's no extra function calls for most of the codecs. `_DECODER` and the few helpers are outside the function:
```python
def _decodeNull(msg):
    frame = msg.next_frame()
    size = FixSizedTypesCodec.decode_int(frame.buf, 0)
    return [None] * size

def _decodeVarChar(msg):
  return ListMultiFrameCodec.decode_contains_nullable(msg, StringCodec.decode)

_DECODER = {
    SqlColumnType.VARCHAR: _decodeVarChar,
    SqlColumnType.BOOLEAN: ListCNBooleanCodec.decode,
    # ...
    SqlColumnType.NULL: _decodeNull,
}
```

I think at the very least we should reduce the number of `column.append`s:
```python
if column_type_id == SqlColumnType.VARCHAR:
    column = ListMultiFrameCodec.decode_contains_nullable(msg, StringCodec.decode)
elif column_type_id == SqlColumnType.BOOLEAN:
    column = ListCNBooleanCodec.decode(msg)
# ...
columns.append(column)
```